In [1]:
import csv
import xlrd
import numpy as np
import pandas as pd

# https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )


In [2]:
# annual_geographical_mobility_rates_type_movement
annual_geographical_mobility_rates_type_movement.sample(3)

Table with row headers in column A and column headers in rows 5 through 7.  \
173                                         Footnotes:                           
93                                           2016-2017                           
123                                                NaN                           
164                                          1954-1955                           
61                                                 NaN                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
173        NaN        NaN        NaN        NaN        NaN        NaN   
93         100         89         11       10.6        6.8        3.8   
123        NaN        NaN        NaN        NaN        NaN        NaN   
164        100       79.6       20.4       19.9       13.3        6.6   
61         NaN        NaN        NaN        NaN        NaN        NaN   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
173        NaN        NaN        NaN  
93         2.1        1.7        0.4  
123        NaN        NaN        NaN  
164        3.5        3.1        0.6  
61         NaN        NaN        NaN

annual_geographical_mobility_rates_type_movement

notes:
    seems to cover back to 1947
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [3]:
# annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.sample(3)

Table with row headers in column A and column headers in row 5.  \
142                                .Movers from abroad                
175                                       .Outmigrants                
260                                        .Inmigrants                
233                                       .Outmigrants                
128                                .Movers from abroad                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  
142        337        254        600        565  
175        708        753       1105        806  
260        482        842       1329        994  
233        714        703       1071        768  
128        161        179        550        380

annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions

notes:
    seems to cover back to 1980
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [4]:
# inmigration_outmigration__net_migration_metropolitan_status
inmigration_outmigration__net_migration_metropolitan_status.sample(3)

Table with row headers in column A and column headers in row 5.  \
169                                          1989-1990                
103                                   ..Central cities                
72                                                 NaN                
142                                                NaN                
151                                          1992-1993                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
169        NaN        NaN        NaN        NaN        NaN  
103       3804       6282    -2,477*        669    -1,809*  
72         NaN        NaN        NaN        NaN        NaN  
142        NaN        NaN        NaN        NaN        NaN  
151        NaN        NaN        NaN        NaN        NaN

inmigration_outmigration__net_migration_metropolitan_status

notes:
    seems to cover back to 1985
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [5]:
# geographical_mobility_by_tenure
geographical_mobility_by_tenure.sample(3)

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
82                                                 NaN                                                                                                                                 
314                          ..In owner-occupied units                                                                                                                                 
160                                          1990-1991                                                                                                                                 
436                                          2003-2004                                                                                                                                 
441                                          2002-2003                                                                                                                                 

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
82         NaN        NaN        NaN        NaN        NaN        NaN   
314        100       91.8        8.2        8.1        5.1        2.9   
160        NaN        NaN        NaN        NaN        NaN        NaN   
436        NaN        NaN        NaN        NaN        NaN        NaN   
441        NaN        NaN        NaN        NaN        NaN        NaN   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
82         NaN        NaN        NaN  
314        1.8        1.1        0.2  
160        NaN        NaN        NaN  
436        NaN        NaN        NaN  
441        NaN        NaN        NaN

geographical_mobility_by_tenure

notes:
    seems to cover back to 1987
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [6]:
# reason_for_move_specific_categories
reason_for_move_specific_categories.sample(3)

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
64  *The foreclosure/eviction reason for move was ...                                                                                    
21                                          2006-2007                                                                                    
7                                         2017-2018/1                                                                                    
26                                          2002-2003                                                                                    
25                                          2003-2004                                                                                    

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
64        NaN        NaN        NaN        NaN        NaN        NaN   
21      38681       2296       3808       5555       3789        648   
7       32352       1435       4073       3583       3341        492   
26      40093       2679       2814       5055       3546        749   
25      38995       2401       2716       4357       3569        922   

   Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 12 Unnamed: 13 Unnamed: 14  \
64        NaN        NaN        NaN  ...         NaN         NaN         NaN   
21       1858        214       1551  ...        6096        2130        3089   
7        1821        348        366  ...        5300         844        2551   
26       1275        101        576  ...        7942        1530        2622   
25       1444        128        561  ...        8235        1840        2865   

   Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  \
64         NaN         NaN         NaN         NaN         NaN         NaN   
21           -        2628         743         146         531         177   
7          215        2156         903         215         579         174   
26           -        4406        1010         160         565           -   
25           -        4009        1118         224         394           -   

   Unnamed: 21  
64         NaN  
21        1130  
7         1605  
26         987  
25         583  

[5 rows x 22 columns]

reason_for_move_specific_categories

notes:
    seems to cover back to 1998
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [7]:
# distance_of_intercounty_move
distance_of_intercounty_move.sample(3)

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
41                        2009-2010 (2010 controls)/2                                                                     
49                                        2003-2004/4                                                                     
28                                            PERCENT                                                                     
51                                         Footnotes:                                                                     
7                                           2016-2017                                                                     

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
41        100         39       21.3       14.4       25.2  
49        100    30.7495    24.5336    21.0006    23.7163  
28        NaN        NaN        NaN        NaN        NaN  
51        NaN        NaN        NaN        NaN        NaN  
7       12033       4781       2414       1788       3051

distance_of_intercounty_move

notes:
    - seems to cover back to 2002
    - needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

## Aggregate Dates
    - first recording year varies by up to 55 years 
| first recording year | dataset |
|------|------|
| 1947 | annual_geographical_mobility_rates_type_movement |
|------|------|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|------|------|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1998 | reason_for_move_specific_categories |
|------|------|
| 2002 | distance_of_intercounty_move |

- let's drop the rows clustered with NaNs

In [8]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
# df
df.sample(7)

Table with row headers in column A and column headers in rows 5 through 7.  \
174  1/ The 2014 CPS ASEC included redesigned quest...                           
150                                          1965-1966                           
148                                          1967-1968                           
45                                           1990-1991                           
66                                           1965-1966                           
12                                           2014-2015                           
121                                          1996-1997                           
5                                                  NaN                           
93                                           2016-2017                           
87                                           1948-1949                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
174        NaN        NaN        NaN        NaN        NaN        NaN   
150        100       80.2       19.8       19.3       12.7        6.6   
148        100       80.5       19.5       18.8       11.8          7   
45      244884     203345      41539      40154      25151      15003   
66      190242     152656      37586      36703      24165      12538   
12      312295     275971      36324      34651      22905      11746   
121        100       83.5       16.5         16       10.5        5.5   
5          NaN        NaN        NaN        NaN        NaN      Total   
93         100         89         11       10.6        6.8        3.8   
87      144101     116498      27603      27127      18792       8335   

     Unnamed: 7       Unnamed: 8 Unnamed: 9  
174         NaN              NaN        NaN  
150         3.3              3.3        0.5  
148         3.4              3.6        0.7  
45         7881             7122       1385  
66         6275             6263        883  
12         6653             5093       1673  
121           3              2.4        0.5  
5    Same state  Different state        NaN  
93          2.1              1.7        0.4  
87         3992             4344        476

notes:
    - 156 rows remain
    - some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    - row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    - rows 173 - 185 contain only 1 NaN value

actions:
    - drop all rows without 2 or more non-NaN values
    - merge rows 3 , 4 , and 5 
        taking out 'Total'

In [9]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement.copy()

# drop columns with less than 2 non-NaN values
df = df.dropna( thresh= ( 2 ) )

# collect column names as pandas series
col_names_0to4_9 = df.iloc[ 0 ]
col_names_5to6 = df.iloc[ 1 ]
col_names7to8 = df.iloc[ 2 ]

# initialize list for column names
column_names = []

# fill list with first row of column names , no exceptions
for column_name in col_names_0to4_9:    
    column_names.append( column_name )

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 5 and 6 in column_names with column names 5 and 6
for column_name in col_names_5to6:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 7 and 8 in column_names with column names 7 and 8
for column_name in col_names7to8:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# reset column names
df.columns = column_names

# drop rows used to extract column names
df = df[3:]

# reset dataframe index
df = df.reset_index( drop=True ) 

# let's see
df.sample(7)

Mobility period Total, 1 year old and over Same residence (non-movers)  \
24        1997-1998                     265209                      222702   
107       1983-1984                        100                        82.7   
42        1980-1981                     221641                      183442   
118       1964-1965                        100                        79.3   
12        2007-2008                     294851                      259685   
65        1949-1950                     146864                      118849   
96        1993-1994                        100                        83.3   
91        1998-1999                        100                        84.1   
68        2017-2018                        100                        89.9   
47        1967-1968                     194621                      156735   

    Total movers Different residence in the United States Same county  \
24         42507                                    41304       27082   
107         17.3                                     16.8        10.4   
42         38200                                    36887       23097   
118         20.7                                     20.1        13.4   
12         35167                                    34022       23013   
65         28015                                    27526       19276   
96          16.7                                     16.3        10.4   
91          15.9                                     15.4         9.4   
68          10.1                                      9.8         6.2   
47         37886                                    36603       22960   

    Different county Same state Different state Movers from abroad  
24             14222       7867            6355               1203  
107              6.4        3.6             2.8                0.5  
42             13789       7614            6175               1313  
118              6.8        3.5             3.3                0.5  
12             11009       6282            4728               1145  
65              8250       4360            3889                491  
96               5.8        3.2             2.6                0.5  
91               5.9        3.1             2.8                0.5  
68               3.6          2             1.5                0.4  
47             13643       6607            7035               1283

as of now, we have:
    1) removed rows with NaN values
    2) extracted, sorted, and set column names
    
next:
    - we need to clean up the 'Mobility period' column
        - samples from current:
            2017-2018
            2013-2014 (98,000 address file)/1
            2011-2012
            2010-2011 (2010 controls)/2
            2000-2001 (SCHIP, 2000 controls)/5
            1970-1971/8
        - what we want:
            2018 << 2017-2018
            2014 << 2013-2014 (98,000 address file)/1
                requires further investigation
            2012 << 2011-2012
            2011 << 2010-2011 (2010 controls)/2
                requires further investigation
            2001 << 2000-2001 (SCHIP, 2000 controls)/5
                requires further investigation
            1971 << 1970-1971/8

In [10]:
# determine type of data in 'Mobility period' 
def test_type( dataframe , test_type ):
    '''
    input) dataframe w/ specified column 
        - e.g. df['specified column']
    input) type wanting to test
        - e.g. str 
    
    output) count of instances which are not type == test_type
    '''
    non_test_type = 0
    for _ in dataframe:
        if isinstance( _ , test_type ) == False:
            non_test_type += 1
    return non_test_type


test_000 = test_type( dataframe=df[ 'Mobility period' ] , test_type=str )
test_000

0

as test_000 returns 0:
    - the type of all datapoints in 'Mobility period' is str (string)
    - we will not need to change any points to have uniformity 
    
next up:
    - clear low-hanging fruit
    - change the simple ranges to int for set year
        - e.g. change "2017-2018" to "2018"

In [11]:
# identify changes which can be implimented without further investigation
mobility_period = df[ 'Mobility period' ]
simple_year_ranges_count_ = 0
confirmed_simple_year_ranges_count_ = 0

# count how many simple-fix years we have
for datapoint in mobility_period:
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # pretty sure we know what's going on but will investigate further         
        simple_year_ranges_count_ += 1
        # first 4 characters of string , can be converted to int
        first_year = int( datapoint[ :4 ] )
        # last 4 characters of string , can be converted to int
        second_year = int( datapoint[ -4: ] )
        # first_year and last_year are post-1940 (because expected), and year range is 1 year 
        if first_year and second_year > 1940 and second_year - first_year == 1:
            # the datapoint is within our risk threshold
            confirmed_simple_year_ranges_count_ += 1

# if everything we thought was what we think it is 
if simple_year_ranges_count_ == confirmed_simple_year_ranges_count_:
    print(f'total: {confirmed_simple_year_ranges_count_}\nratio: {confirmed_simple_year_ranges_count_/len(mobility_period)}')

total: 110
ratio: 0.8088235294117647


notes:
    - know count of lowest fruit to be 110 ( > 80.88% , not bad)
    - just confirmed status of 110 points to be (at least) similar enough to extract years
    - knowing simple_year_ranges_count_ allows the conformation of outcome via count compairson
    
actions:
    - convert those 110 year ranges to years represented by the 2nd year in each range

In [12]:
less_low_mobility_periods = []
# so row 0 == row 0
_ = -1
# little bit of ctrl + c , little bit of ctrl + v
for datapoint in mobility_period:
    _ += 1
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # replace datapoint w/ int of last 4 characters of string 
        mobility_period[ _ ] = int( datapoint[ -4: ] )
    else:
        # keep what's already there
        less_low_mobility_periods.append( datapoint )

mobility_period.sample(7)

20     2000-2001 (SCHIP, 2000 controls)/5
58                                   1957
10            2009-2010 (2000 controls)/3
112                           1970-1971/8
114                                  1969
108                                  1983
29            1992-1993 (1990 controls)/6
62                                   1953
61                                   1954
116                                  1967
Name: Mobility period, dtype: object

- we have converted the low hanging fruit